### ISSUE TITLE
Standardize version number generation

### ISSUE THREAD

**Issue by CharliePoole** [nunit/NUnit.System.Linq#8]
The versions now being generated by GitVersion for NUnit.System.Linq follow GitVersion's internal standards, which are not the same as our own.

The latest master build would have been something like 0.6.1-CI-23 by our current standards. However, the GitVersion generated numbering is 0.6.1+23.build.62 and the package published to MyGet is version 0.6.1. I'm assuming the internal AssemblyVersion is also 0.6.1 as we can't actually tell any longer without examining the assembly.

All of this is OK and expected at this point. NUnit.System.Linq was chosen for us to work with GitVersion because it is only used by us. However, we should conclude our experiments quickly and remove any extraneous packages form MyGet.

This issue will be used to decide on our versioning going forward and will also apply to other packages.

@asbjornu I'll get this started by creating a branch and giving you access to it.

**Comment by CharliePoole** [comment]
I created the issue-8 branch and added @asbjornu as a collaborator for this repo. it seems to me that the first step is to generate the configuration so we can see it. Right?

**Comment by CharliePoole** [comment]
Oh... I see you're in Oslo and probably asleep now. :-) When you're available, would you generate the configuration and push to the branch? Then we can figure out together with other developers on the team how we may want to modify it.

I'm currently working on the NUnit 3 TestAdapter 3.4 release. Once that's out, I want to start following in your footsteps to get this implemented in that project.

**Comment by asbjornu** [comment]
@CharliePoole Just to get this right, what you'd like to replicate with GitVersion are these lines that were removed from `build.cake`?

``` c#
var tag = AppVeyor.Environment.Repository.Tag;      

if (tag.IsTag)      
{       
    packageVersion = tag.Name;      
}       
else        
{       
    var buildNumber = AppVeyor.Environment.Build.Number;        
    packageVersion = version + "-CI-" + buildNumber + dbgSuffix;        
    if (AppVeyor.Environment.PullRequest.IsPullRequest)     
        packageVersion += "-PR-" + AppVeyor.Environment.PullRequest.Number;     
    else if (AppVeyor.Environment.Repository.Branch.StartsWith("release", StringComparison.OrdinalIgnoreCase))      
        packageVersion += "-PRE-" + buildNumber;        
    else        
        packageVersion += "-" + AppVeyor.Environment.Repository.Branch;     
}
```

**Comment by CharliePoole** [comment]
No...but that code represents our existing branch naming policy. The new policy in this repo is "whatever invisible thing GitVersion is doing" :-)

Whatever we do in this repo will be ported to **all** of our NUnit projects. Those projects have public feeds - as opposed to this one, which is publicly accessible but not advertised or used by anyone except us. Since we almost never expect to change this repo, it's a good one for experiments, but if we had done the same thing in the nunit project, for example, all of our feeds would now be broken.

So we need to 
1. Make the configuration visible
2. Decide what we want the new standard to be
3. Implement it in this repo
4. Port it to all our other repos

In an earlier comment, you said that you could create a config file that we would then change to set standards for versioning, including per-branch standards. Creating it would be step 1, which is what I was asking you to do. Right now, looking at the generated versions we don't even know what they mean. Some of them contain two int numbers, for example. How does the programmer know where they came from? How does the user know what they are if we publish them? Which ones do we want to deploy? So it seems to me that visibility is the first step to getting anywhere with this.

If you could easily add comments or even change it to replicate the existing standard, that would be great, but you shouldn't spend a lot of effort on that because it's pretty certain that some parts of the existing standard will be changed in step 2 anyway because GitVersion gives us better capabilities.

This Issue is going to be more about discussion with the team than coding. :-)

**Comment by asbjornu** [comment]
@CharliePoole I see. I've committed and pushed a `GitVersion.yml` file with the defaults, so we can now go on and adjust them. I'm just not entirely sure what I'm supposed to adjust. :smile:

**Comment by CharliePoole** [comment]
That's why we're pairing on this issue. Anyway, I'm not sure either, but I expect we'll get sure as we proceed. I'll take a look and post any questions for you, then I'll start to invite other people to the party.

**Comment by asbjornu** [comment]
@CharliePoole Excellent! :+1:

**Comment by CharliePoole** [comment]
OK... First question... it seems like the way we work is more like ContinuousDeployment than ContinuousDelivery. Can you tell me (or point me to info about) how GitVersion acts differently using the two modes?

**Comment by CharliePoole** [comment]
Never mind... I got the terms switched in my brain... I'm used to "deployment" meaning not yet in production but ready for it but the modern usage is that deployment == to production. If you work long enough, the terms don't just change they even reverse meanings!

**Comment by CharliePoole** [comment]
I found the reference doc, which is helping a bit.

**Comment by asbjornu** [comment]
@CharliePoole Yea, the [documentation](http://gitversion.readthedocs.io/en/latest/) is quite extensive. Glad you figured it out, although I agree that the terms "deployment" and "delivery" are ambiguous, which makes them hard to distinguish.

**Comment by CharliePoole** [comment]
OK, another couple of questions...

What if the branch doesn't match any of the patterns? 

What if you aren't on a branch at all - i.e. you just took an export of the code from git?

**Comment by CharliePoole** [comment]
@nunit/core-team @nunit/vs-extensions-team @nunit/contributors 

I'm using this issue, as a vehicle to figure out how we should be doing versioning across all our projects. Please read the comments and tune in to what is discussed here as it will impact a lot of stuff.

Questions to be sorted out, eventually...

Should every build use a unique version? file virsion? informational version?

Which builds should be on feeds? How many feeds? How versioned?

Finally - what exact naming pattern will we use for each kind of package we build.

All of this in the context of what GitVersion is capable of.

**Comment by asbjornu** [comment]
@CharliePoole:

> What if the branch doesn't match any of the patterns?

The default naming strategy is to include the name of the branch. This is expressed through the `tag` configuration property, which has `useBranchName` as the default for all branches except known git flow branches such as:
- `master`
- `develop`
-  `hotfix`
- `support`
- `release`
- `pull-request`

These known branches have default as can be seen in the `GitVersion.yml` file.

> What if you aren't on a branch at all - i.e. you just took an export of the code from git?

Then we have a problem. It can be addressed by Cake, for instance, but without the `.git` folder, GitVersion is clueless.

**Comment by CharliePoole** [comment]
Is there an alternative where we keep version info for use in such builds in AssemblyInfo.cs but write and use a modified copy for the builds where GitVersion comes into play? Alternatively, is there a way to tell GitVersion "If running outside of a repository, use this version as the default?"

**Comment by CharliePoole** [comment]
To be clear, I'm just gathering info here. I'm not sure this is really that great a problem any longer, now that we use Git. In the far past, with CVS, there was a much greater need to use exported source. 

@nunit/core-team What do you think? Is it OK that our AssemblyInfo.cs contains no version info but that it gets filled in by GitVersion when it's being built? Come to think of it, should we be packaging source at all these days, when people can just clone our repository? As always in this issue, I'm talking about all our projects, not just NUnit.System.Linq.

**Comment by rprouse** [comment]
I've been questioning the need to package source lately, especially since GitHub creates a ZIP download whenever we tag a release. I am also fine with the version getting filled in at compile time, it is actually closer to where I wanted to get. It would be nice if there was only one authoritative version for each repository.

I've never tried it, but do you get a default version if not is specified and GitVersion hasn't injected one? In other words, is `AssemblyVersion` or `FileVersion` required to compile? Does GitVersion only inject the version, or can it overwrite one if present? If so, we could put something like 9.9.9.9 into the various AssemblyInfos to indicate it is an unofficial build.

**Comment by OsirisTerje** [comment]
I'm fine with gitversion handling this.  The assemblyinfo files should then stay with 1.0.0.0, and prod would almost always be higher (!).  @rprouse : having it at 9.9.9.9 means higher than most of what we have, I would like to see it lower, so that prod versions always takes precedence.  
Source is on github, so fully agree it is not needed in release packages.  A link is more than enough.

About feeds:  one production feed (nuget.org), and a prerelease feed on myget would be nice. Many have nightly feeds, but I think a prerelease feed is more important.  Anyone can build from sources if they want the latest. 

Build numbers would be updated on each build, the latest being the the incrementally updated number.  Versions should be identical, except for the last part, but pre-releaseinfo should be kept.  

And just throwing in: 
We talked earlier of an odd- even practice on build numbers, to separate out release versions from pre-release and things-in-process.   nuget would not need that, since it has the letters (1.2.3-beta1) to indicate pre-releases, but anything else would need something else. 
I am not sure how gitversion handles this.....
Packages should follow closely to the build numbers,  if the build is a pre-release, the package should be marked as such.

**Comment by CharliePoole** [comment]
Easy decisions first: let's stop packaging source then! I'll make the change right away in this one and in the NUnit 3 Test Adapter, which has an outstanding PR. We can drop source in others as it's convenient. 

AssemblyVersion: As currently implemented by @asbjornu we don't have any version in the AssemblyInfo for this repo. To implement that, he used the GitVersion msbuild task, which will execute whenever anybody builds this project, following the rules we establish for versioning. The assemblyinfo in GitHub is replaced by a generated one that adds assembly and file versions.

I was dubious but became convinced that it's a good approach. There will be no version info in any of our files, therefore we will have nothing to update. Builds on local systems will do the same thing as builds on AppVeyor or Travis. Still, I admit that not having it in the file makes me nervous. I'm hoping that getting more familiar with GitVersion will make me more comfortable. Right now, I don't understand enough to know what version it is that I'm about to build!

I have asked @asbjornu if there is an alternate way that this can work where we continue to have a fixed version in the AssemblyInfo. Waiting to hear back.

@rprouse I don't know what happens if you happen to build without specifying a version anywhere. I'll have to do an experiment on that. Still... in principle this would only happen if somebody did an export and made a non-git directory, so it's very much an edge case.

@OsirisTerje Yes, two feeds is what I think as well. The way I have implemented it here, only merged builds of master go to the MyGet feed. Rob and I have also talked about using the account feed on AppVeyor instead of the Project feed as we now do. It seems we are probably misusing the project feed and it suffers from the fact that we can't remove anything!

@OsirisTerje Could you clarify what you mean by incrementing the build number? The term is used in a few different ways: part of the standard assembly version, the AppVeyor build number, the build number that GitVersion tacks on as metdadata!

One issue we will need to deal with... vsix packages can't use a prerelease suffix internally. We'll cross that bridge when I try to apply GitVersion to the NUnit 3 Adapter.

The odd/even convention is handly when long-term development is going on because people can choose whether to get the latest stable version or the development version. However, what I didn't think of originally is that implementing it really requires separate feeds as well. People would need a place to go to get the development version. The downside of this is more work for us. The upside is that you can do less frequent production releases, since folks who need the latest have a place to get it. Of course, they can also get that from a separate pre-release feed on MyGet. We should talk about that and make any changes rather quickly if we are going to make them. It could be we want to do different things with different projects as well.

I think we'll have to set up GitVersion in our case so that it never increments the major or minor component automatically. If we are following semver - or something close to it - then only we know if a particular change is a new feature or a breaking change. We could do minor increments automatically, but that would mean being quite disciplined in how we distinguish bug fix branches from feature branches. I'm not sure if we can do that. :-)

**Comment by asbjornu** [comment]
@CharliePoole:

> Is there an alternative where we keep version info for use in such builds in AssemblyInfo.cs but write and use a modified copy for the builds where GitVersion comes into play?

Not with GitVersionTask. Then we would need to use `GitVersion.exe`, which doesn't (really) work compile-time, but on the upside supports patching of existing `AssemblyInfo.cs` files.

> Alternatively, is there a way to tell GitVersion "If running outside of a repository, use this version as the default?"

We can hard-code `next-version` in `GitVersion.yml`, although I'm not sure GitVersion will actually execute at all with the `.git` directory missing. If that's a desired use-case, I'll have to test it to make sure it actually works.

> I don't know what happens if you happen to build without specifying a version anywhere.

I'm pretty sure that without any `Assembly*Version` attributes, the version number defaults to `0.0.0.0`.

@OsirisTerje:

> We talked earlier of an odd- even practice on build numbers, to separate out release versions from pre-release and things-in-process. nuget would not need that, since it has the letters (1.2.3-beta1) to indicate pre-releases, but anything else would need something else.  I am not sure how gitversion handles this.....

I would **strongly** advice against doing odd/even versioning, both because users find it hard to understand and because it's incompatible with [semver](http://semver.org/) (which is what most people seem to have the best intuitive understanding of these days, with penetration and understanding increasing every day). People seem to be [moving away from odd/even versioning](https://nodesource.com/blog/nodejs-is-semver/), so I find it peculiar that NUnit would start employing that scheme now.

More importantly, I don't see that odd/even versioning is at all necessary, given the availability of a pre-release MyGet feed. For long-term support of a certain unstable branch of `2.0`, just maintain that in a `support/2.x-unstable` and have packages from it published to an unstable feed. I don't really see the use-case of having long-term support for unstable packages, though.

@CharliePoole:

> The odd/even convention is handly when long-term development is going on because people can choose whether to get the latest stable version or the development version.

They can do that with semver too. With MyGet, you can even maintain a different feed per branch, to make consumption easy. I don't really see why you would want long-term support for an outdated unstable branch, though.

If you have ongoing unstable work for both version 2 and 3 of NUnit, they will have a different major version number, which will separate them in the feed. If people [configure their dependencies correctly with version intervals](https://docs.nuget.org/Create/Versioning), they won't accidentally upgrade from version 2 to 3, but instead follow version 2's unstable/stable ticks and tocks for as long as that branch is being maintained.

> I think we'll have to set up GitVersion in our case so that it never increments the major or minor component automatically.

That can be configured. The default is to increment `patch` for most branches, but for `develop` it's `minor` because you usually want the latest release from `develop` to be of a higher version number than any previous releases from any other branch originating from the same tag. This is important when sorting packages in NuGet. If you screw this up, you'll end up with a pre-release package that suddenly is considered "latest" that shouldn't and you will have to remove it from the feed to fix it.

> If we are following semver - or something close to it - then only we know if a particular change is a new feature or a breaking change.

Yes, absolutely. Any merge against `develop` or `master` is worthy of an explicit version number increment. [How you do the increment, however, is up to you](http://gitversion.readthedocs.io/en/latest/more-info/version-increments/). You can do it with tagging (my preference, since it's more explicit and easier to figure out after the fact) or by adding `+semver: minor` to your commit messages, for instance.

> We could do minor increments automatically, but that would mean being quite disciplined in how we distinguish bug fix branches from feature branches. 

Packages produced from bugfix and feature branches will have a different pre-release tag in their version number, so they will be easily distinguishable. They will, however, be considered "the same" by NuGet, so it might suggest upgrades from a bugfix to a feature branch package if the latter has a higher semver, though. Problems like this can be avoided by using different feeds, but that of course adds administrative overhead.

I wish NuGet had a "branch" concept, so newer versions of a package from another branch than the one currently installed weren't suggested as upgrades unless explicitly asking NuGet to suggest them.

**Comment by CharliePoole** [comment]
@asbjornu Thanks for the answers. I think I'll experiment with the Cake tool, which is essentially a wrapper around GitVersion executable. I'll do that for the NUnit3 Test Adapter, which is the next thing I want to apply GitVersion to - issue nunit/nunit3-vs-adapter#199. Then we can compare the two approaches and see what we like best.

**Comment by CharliePoole** [comment]
@asbjornu Regarding the odd/even convention, understand that it's not a proposal - it's what we have been doing since 3.0. That's why 3.2 and 3.4 were the next releases after 3.0 for both NUnit and the NUnit3 Test Adapter. So, if we decide to drop it we will next have to decide _how_ to drop it, i.e. if we need to phase it out. Possibly, if we made that decision right away there wouldn't need to be any phasing. I just wanted you to be up to speed on that before getting to the meat of the issue.

I'm the big advocate for odd / even and I provided the push for implementing it with 3.0. My main reason is that we go a long time between feature releases and some people want to get those features early. The odd / even convention is a long-time Linux convention and I agree that it's a bit confusing to Windows people. (Even though .NET is primarily for Windows, the development of this framework, and the whole testing/TDD movement were long dominated by non-Windows, or at least non-MS people.) That's history, however.

Yes, I am seeing that the ability to have different feeds  would essentially eliminate most of the problem that odd / even was intended to resolve. I was imagining a single feed, with people who did not want to "upgrade" from stable to unstable having to use nuget version bracketing to prevent the next odd version from being installed. They would also have to explicitly upgrade to the next even version, which is an inconvenience. I also thought that we would have incompatible pre-release suffixes for the odd and the even versions. That is, odd would use our CI suffixes, while even would use Alpha, Beta, RC. 

I can see that having different feeds - say NUnit stable and unstable - on MyGet would provide the same functionality and possibly be simpler. What do the rest of you think?

**Comment by CharliePoole** [comment]
@All Does this discussion open up the possible need to change our branching model? Currently we have master, issue branches and (pre)release branches. 

We don't distinguish bugs from features in naming the branches. We could do that, but as I said earlier, it would take a bit of discipline. For our own work, we generally label issues as bugs, features or enhancements, so it's possible we could just start creating branches with different names in order to give GitVersion something to base its decisions on. I'm not sure how we would handle PRs from contributors though.

We earlier talked about GitFlow and decided not to go with it. One advantage it would have is that the dev branch could have a different set of versioning rules from master. I'm not advocating for this, just throwing out the possibility.

It seems like we are at a point, just after a release, where we could make changes easily so lets consider the options. One thing - even though it's not absolutely necessary, I'd like to see us adopt one set of standards across all the projects.

**Comment by rprouse** [comment]
> I would strongly advice against doing odd/even versioning, both because users find it hard to understand and because it's incompatible with semver (which is what most people seem to have the best intuitive understanding of these days, with penetration and understanding increasing every day). People seem to be moving away from odd/even versioning, so I find it peculiar that NUnit would start employing that scheme now.
> 
> More importantly, I don't see that odd/even versioning is at all necessary, given the availability of a pre-release MyGet feed. For long-term support of a certain unstable branch of 2.0, just maintain that in a support/2.x-unstable and have packages from it published to an unstable feed. I don't really see the use-case of having long-term support for unstable packages, though.

I also dislike the odd/even versioning for the same reasons. We never release odd numbered builds in any format other than NuGet feeds, so I don't think the convention adds anything for us that the pre-release nuget suffixes also add in a way that is more familiar with the majority of our users.

**Comment by rprouse** [comment]
> We don't distinguish bugs from features in naming the branches. We could do that, but as I said earlier, it would take a bit of discipline. For our own work, we generally label issues as bugs, features or enhancements, so it's possible we could just start creating branches with different names in order to give GitVersion something to base its decisions on. I'm not sure how we would handle PRs from contributors though.
> 
> We earlier talked about GitFlow and decided not to go with it. One advantage it would have is that the dev branch could have a different set of versioning rules from master. I'm not advocating for this, just throwing out the possibility.

Other than the fact that we made the mistake of branching off master instead of the release branch for the `3.4.1` hotfix, we have a branching model that is working fairly well and fits in with GitHub well, so I would prefer not to change it unless we have strong reasons to do so.

As for branch naming, I'd be fine with switching to a new naming convention, as long as it isn't too complicated. I want to keep it as simple as possible for new developers to on-board without reading reams of documentation on our development processes :smile:

**Comment by asbjornu** [comment]
@CharliePoole:

> Regarding the odd/even convention, understand that it's not a proposal - it's what we have been doing since 3.0. 

Oh, I wasn't aware of that. I've been happily upgrading to every new version as if it followed semver. Although I understand the history behind this, I disagree with the conclusion. NuGet is the prevalent distribution model and package manager for .NET, NUnit is written for a .NET audience and NuGet uses semver. Anything breaking the rules of semver practically breaks NuGet and consumers of the package.

@rprouse:

> As for branch naming, I'd be fine with switching to a new naming convention, as long as it isn't too complicated.

Why not just use [git-flow](http://jeffkreeftmeijer.com/2010/why-arent-you-using-git-flow/) directly and its rather standardized naming scheme:
- `master`
- `develop`
- `feature/*`
- `release/*`
- `hotfix/*`
- `support/*`

That has the added benefit of being a known scheme to GitVersion, which then won't need any special configuration to figure out everything and do its thing. It's also a scheme well known and implemented in most visual Git tools, such as [SmartGit](https://www.syntevo.com/doc/display/SG/Git-Flow) and [Tower](https://www.git-tower.com/help/mac/integration/git-flow), all of which makes it the most known naming scheme for Git branches in existence, I assume.

**Comment by rprouse** [comment]
> Why not just use git-flow directly and its rather standardized naming scheme

Mainly because we decided to use GitHub flow and don't feel a strong need for a develop branch :smile: 

We do use some of the GitFlow branches like `release/*` and I wouldn't mind switching to `feature/*` instead of `issue-xx`. I just like that developers can pull master, build and test the latest fixes. Then again, we could always just switch our default branch on GitHub to `develop` and treat it like we do master. I've just never really seen the point of the master branch in GitFlow since it is just tags of release points. I just tag the release branches. GitFlow is probably a bit cleaner.

You also say it is supported by most tools which is true, but it isn't supported by GitHub for Windows or GitHub for Visual Studio AFAIK which GitHub does a pretty good job of pushing on users. I don't use them and would actually find Git Flow easier in many ways, but I do worry about junior developers.

Don't take my comments too negatively. I actually like GitFlow, I just want to keep things as simple as possible for NUnit and only add complexity if it is needed. I also feel that more developers are comfortable with the GitHub Flow model. If people really want to change though, it won't take much to convince me :wink:

**Comment by OsirisTerje** [comment]
There is also a gitflow extension for Visual Studio which works well, https://visualstudiogallery.msdn.microsoft.com/f5ae0a1d-005f-4a09-a19c-3f46ff30400a  

The only thing I miss is the Issue branch "folder", it is not features and it is not hotfixes, and a lot of what is done is bug fixes, so......

I agree with @rprouse about keeping things simple, and in our case I dont really see the need for the develop branch, but heck, it is just a pointer, we dont really need to bother too much.

**Comment by CharliePoole** [comment]
I had already come to see that the odd / even versioning scheme wasn't getting me what I had hoped to achieve with it. So we either need to make it work or do something different. Let me try to list what it was that I was trying to accomplish...
1. We should be free to add features and try them out without burdening users.
2. Users who don't want to change until we do a formal release should be able to do so readily. 
3. Users who only want to get bug fixes should be able to do that easily.
4. Users who want to follow us closely as new features get added should be able to do that. They should have the choice of doing it by getting binary drops or forking/cloning our repo.
5. When users report a problem, they should be able to tell us exactly what they are using.

Doing all that requires a clear separation between stable and unstable releases and stable and unstable source. It's clear that just using odd and even releases has not done that. It seems that they worked well in the past when the releases were encapsulated in branches - in SVN for example.

So, I'd be ready to drop the odd/even thing but would still want to reach those goals listed.

Users need to get packages from us. We need to have a stable and an unstable feed. The stable feed would contain packages like 3.4.0, 3.4.1, etc. I think that's what nuget.org is for nuget packages. The unstable would be packages like 3.5.0-xxx, 3.5.0-yyy, etc. where the suffix values have yet to be determined. Our myget feed does that for nuget packages and we can create multiple myget feeds if we need to. Does this make sense? If so, how should it tie back to branches?

**Comment by CharliePoole** [comment]
Just for the record, I'm unswayed by the existence of gui tools that support a particular flow. I'm a command-line guy and I would hope that anyone contributing to NUnit would be reasonably comfortable on the command-line. So anything we use needs to have a straightforward implementation in commands and/or aliases as well.

**Comment by asbjornu** [comment]
@rprouse:

> Mainly because we decided to use GitHub flow and don't feel a strong need for a develop branch :smile: 

Well, when you say:

> We do use some of the GitFlow branches like `release/*` and I wouldn't mind switching to `feature/*` instead of `issue-xx`.

…then you're not really using GitHub Flow, but GitFlow.

> I just like that developers can pull master, build and test the latest fixes. Then again, we could always just switch our default branch on GitHub to `develop` and treat it like we do master.

Yes, that's the way to do it. If you want unstable releases without an odd/even versioning scheme, why would you **not** want a `develop` branch? I have great experience with setting `develop` as the default branch on GitHub; it makes it easier on everyone, including outside contributors.

> I've just never really seen the point of the master branch in GitFlow since it is just tags of release points. I just tag the release branches. GitFlow is probably a bit cleaner.

I think GitFlow is cleaner, yes. With it, you won't have **any** commits on `master` that isn't a tagged merge-commit. Very easy to follow. If you want to see details, just poke into whatever was merged from `develop`, which again is mostly merges from feature branches or pull requests.

> You also say it is supported by most tools which is true, but it isn't supported by GitHub for Windows or GitHub for Visual Studio AFAIK which GitHub does a pretty good job of pushing on users. I don't use them and would actually find Git Flow easier in many ways, but I do worry about junior developers.

True, but the fact that we use GitFlow isn't something outside contributors really need to be aware of. How we organize our branches is internal and up to us. Outside contributors can't affect that and we can't affect what they do in their repository, but when they submit a pull request, we can decide what to do with it. 

> Don't take my comments too negatively. I actually like GitFlow, I just want to keep things as simple as possible for NUnit and only add complexity if it is needed.

Sure, no problem. I just utterly fail to see the complexity of GitFlow, that's all. :smile: 

> I also feel that more developers are comfortable with the GitHub Flow model. If people really want to change though, it won't take much to convince me :wink: 

I don't see why anyone outside of the NUnit team would even notice what flow we're using. Okay, we have a `develop` branch, but I doubt most contributors would care as that's just a (new) default, just the same way `master` is today.

@OsirisTerje:

> I agree with @rprouse about keeping things simple, and in our case I dont really see the need for the develop branch, but heck, it is just a pointer, we dont really need to bother too much.

If we abandon the odd/even versioning scheme, from which branch are we going to do unstable releases if we don't have a `develop` branch?

@CharliePoole:

> 1. We should be free to add features and try them out without burdening users.
> 2. Users who don't want to change until we do a formal release should be able to do so readily. 
> 3. Users who only want to get bug fixes should be able to do that easily.
> 4. Users who want to follow us closely as new features get added should be able to do that. They should have the choice of doing it by getting binary drops or forking/cloning our repo.
> 5. When users report a problem, they should be able to tell us exactly what they are using.

Yep:
1. `feature/*` published to a separate MyGet feed.
2. Consume the stable releases on NuGet and follow the `master` branch on GitHub.
3. Use semver. If you have versions 3.0.2 and 3.2.0 with the same bug in it and don't want 3.0.x people to upgrade to 3.2.x, just put up a new 3.0.3 and 3.2.1 package. This sounds theoretical to me, since minor increments shouldn't break anything so people should therefore not be afraid of upgrading from a minor to the next, but it's the same principle if it is indeed a breaking change and a new major increment.
4. Consume pre-releases on MyGet (or NuGet, for that matter) and follow the `develop` branch on GitHub.
5. Have them report the `AssemblyInformationalVersion` of the assembly they're consuming. It will contain everything we need to know, including the hash of the commit that was built. If we're using GitVersion, that is.

> Doing all that requires a clear separation between stable and unstable releases and stable and unstable source.

I believe GitFlow and semver gives you this. I don't even thing we need to use MyGet to accomplish the release bit; we can do pre-releases to NuGet too. It only makes it difficult or probably impossible to publish pre-releases from different branches, because NuGet doesn't separate them.

> So, I'd be ready to drop the odd/even thing but would still want to reach those goals listed.

I believe we can reach those with GitFlow and semver.

> Users need to get packages from us. We need to have a stable and an unstable feed. The stable feed would contain packages like 3.4.0, 3.4.1, etc. I think that's what nuget.org is for nuget packages. The unstable would be packages like 3.5.0-xxx, 3.5.0-yyy, etc. where the suffix values have yet to be determined. Our myget feed does that for nuget packages and we can create multiple myget feeds if we need to. Does this make sense?

Yes, this sounds like what I have in mind.

> If so, how should it tie back to branches?

That depends. :smile: The branch/publish matrix might look something like this:

| Branch | Feed | Example version number |
| :-- | :-- | :-- |
| `master` | NuGet stable | `3.4.1` |
| `develop` | NuGet or MyGet pre-release | `3.5.0-alpha.1` |
| `feature/*` | MyGet feature feed | `3.4.2-feature-x.4` |
| `release/*` | MyGet release feed | `3.4.1-beta.8` |
| `hotfix/*` | MyGet hotfix feed | `3.4.1-beta.10` |
| `support/*` | NuGet stable | `2.6.1` |

**Comment by OsirisTerje** [comment]
@asbjornu :  Good points!    Agree.  And yes, without odd/even, develop branch makes even more sense, also your other points on that makes sense.

**Comment by CharliePoole** [comment]
Agree, all good points.

As I wrote earlier, let's get back to basics of what we want to accomplish and how to do it. Discussion of "name brands" of flows is no more helpful than discussion of whether somebody is truly doing XP, Scrum, etc. in the Agile world.

What **we** are doing isn't GitFlow and it isn't GitHubFlow. What we will do needs to be decided bit by bit, not by picking a "brand", which only fuels debate rather than problem solving.

So... here is one piece...

Do we need an unstable feed? I say yes, separate from nuget.org. Our nuget feed only has room for one sequence of pre-releases: like alpha, beta and RC. Additionally, it follows semver. An unstable feed can't really follow semver, at least in spirit since by definition it may add and drop features at any time. If you follow an unstable feed, you follow the ups and downs of our efforts as we add things, try them out and then either keep them or drop them. Such a feed could either have a new release for each merge or just a daily release. I doubt we need both.

You could create such a feed without a dev branch, but it's a lot easier with one and that gives folks building from source a place to get the unstable bits. So I would like to have the branch. My preferred name BTW is dev but I can live with the longer one.

A few years back I had talked about doing this in conjunction with the odd/even split. Master would have been even, dev odd. After discussion we agreed it was too complicated. We dropped dev and kept the split. I think now we should have kept the branch.

Let's find out if we agree on any of the above before moving on to release branches, etc.

**Comment by asbjornu** [comment]
@CharliePoole:

> What we **are** doing isn't GitFlow and it isn't GitHubFlow. What we will do needs to be decided bit by bit, not by picking a "brand", which only fuels debate rather than problem solving.

True, and I agree. However, if you switch from odd/even to semver, you have a need for `develop`, which aligns the NUnit development process pretty straight with GitFlow.

> Do we need an unstable feed?

I agree. From reading your points I even think we need more than one. At least two; one for `develop` and `release/*` (producing "alpha", "beta" etc.) and one for `feature/*` and pull requests (which might be the same thing, depending on how we merge them). The former can (imho, should) be on NuGet, the latter can (imho, should) be on MyGet.

> My preferred name BTW is dev but I can live with the longer one.

My preferred name and what I introduced in my org. before knowing about GitFlow was `development`. We've standardized on `develop` now, though, since that's what "everyone" else is using.

> Let's find out if we agree on any of the above before moving on to release branches, etc.

:+1:

**Comment by CharliePoole** [comment]
Back when I first encountered SCM systems, we used main, test and dev.
Every letter had to be typed, of course. :-)

**Comment by CharliePoole** [comment]
Once we have agreement on having a dev/develop branch and dropping the odd/even split, I'll update the milestones that we set up for nunit and the adapter... 3.6, 3.8, 3.10 will become 3.5, 3.6, 3.7. I don't want to do that till I'm sure we are in agreement.

**Comment by oznetmaster** [comment]
I did not really like odd/even when we last had this discussion, so I am all for the proposed approach.

**Comment by rprouse** [comment]
@asbjornu strong arguments, you are winning me over :+1: 

I am fine with a develop branch. If we are changing, I think we should stick with default GitFlow naming as outlined by @asbjornu above. All tools that I have seen use that as the default for the tool users among us.

![image](https://cloud.githubusercontent.com/assets/493828/16661127/79b5b514-443f-11e6-9166-95f33012f9c2.png)

**Comment by OsirisTerje** [comment]
One small thing more:  We are doing a bunch of bugfixes, intermixed with new feature development.  We have used the "Issue" prefix for that, but since the git flow has a standard name for that too, "bugfix", we should also use that.  See it is not in @asbjornu list above, but it has been added to git flow.

**Comment by CharliePoole** [comment]
I'm OK with adopting GitFlow terminology for whatever we do but I still see value in looking at things one at a time. We don't want to complicate our work by adding a branch type we don't need just because it's in GitFlow.

Note... this is not in response to @OsirisTerje 's bugfix comment... just to the general idea that we might "adopt GitFlow" en masse.

Given what we have decided so far, I just changed our future milestones to increment by 1 rather than 2. NUnit project has 3.5, 3.6, 3.7 scheduled. NUnit 3 Test adapter has 3.5. AFAIK no other projects have established advance milestones.

**Comment by asbjornu** [comment]
@CharliePoole, @OsirisTerje Why isn't the word `hotfix` fitting for your definition of "issue" or "bugfix"? `hotfix/*` is a GitFlow term.

**Comment by CharliePoole** [comment]
@asbjornu Well... I only commented to say I wasn't commenting on it. :-)

However, my understanding is that **some** GitFlow users make a distinction. A bugfix may wait for a release, while a hotfix actually causes an immediate direct release on master and also gets merged to develop. I'm not sure I buy the need for two different naming conventions, but I'm willing to listen. @OsirisTerje ?

I think that there is a big prior decision about how we want to do releases that has to be made before worrying about this detail. How do we handle bugs? That is, if we are on 3.4.0 and we fix a bug, when will that be released to the user. Will it be in 3.4.1 or 3.5.0? Do we do the same thing with all bugs or are critical bugs treated differently. I think the answer depends on how much overhead there is in doing a release. Currently, because of the overhead, we save up fixes until we are ready to release. If we do that for everything we will only have minor releases because there will always be some kinds of features mixed with bugs.

**Comment by asbjornu** [comment]
@CharliePoole Yea, I can see the difference. But as you mention, it depends more on how and when you're able to release than the philosophical semantics of "issue" versus "hotfix" versus "bugfix".

What matters in GitFlow, though, is that it thinks of a `hotfix` branch as something that is created from `master` and then merged back into `master` as well as `develop` to get the fix out there as soon as possible. I'm not sure that's a use-case NUnit needs, but that's how it deals with it. If bugfixes can wait a bit, you can basically just treat bugfixes the same way as `feature` branches.

> That is, if we are on 3.4.0 and we fix a bug, when will that be released to the user. Will it be in 3.4.1 or 3.5.0?

SemVer says it should be 3.4.1, unless it's a breaking change, which would make it 4.0.

> Do we do the same thing with all bugs or are critical bugs treated differently.

Good question. I don't know the history of the NUnit repositories well enough to have a say in that decision. 😃 

> I think the answer depends on how much overhead there is in doing a release. Currently, because of the overhead, we save up fixes until we are ready to release. If we do that for everything we will only have minor releases because there will always be some kinds of features mixed with bugs.

I didn't know there was such an overhead doing releases of NUnit, but I'm happy to let you know that if you adopt GitFlow or a variation of it, you can start using tools like [GitReleaseNotes](https://github.com/GitTools/GitReleaseNotes) to automate the creation of release notes and [GitReleaseManager](https://github.com/GitTools/GitReleaseManager) to help automate the release process.

**Comment by OsirisTerje** [comment]
Bugfix as a term has been added to gitflow, I think it was lately, just because it makes things a bit more clear I assume.  
![image](https://cloud.githubusercontent.com/assets/203432/16670360/23602f22-449a-11e6-9e7f-c8d4bd9ada22.png)

I have just noted that when we work with NUnit we do a bunch of bugfixes, which mostly wait for the next release, and to keep easy track of them I think having them in a bugfix "folder" would be very useful.  Of course it is possible to have them in a "feature" folder, but I can't see any harm separating them out.  When we document releases we often also separate between what is new features and what is bugfixes. 

The use of tools like you mention to help out with release notes and the process sounds great, the more than can be automated the better. (Both your links point to GitReleaseNotes,  the second should have been https://github.com/GitTools/GitReleaseManager I assume).

**Comment by asbjornu** [comment]
@OsirisTerje Ah, look at that. From what branch is `hotfix` normally created from, then? Is it functionally like a `feature` branch, only with a different name? If so, that sounds like a perfect match for NUnit's workflow.

> (Both your links point to GitReleaseNotes, the second should have been https://github.com/GitTools/GitReleaseManager I assume).

Yes, fixed. 😃

**Comment by OsirisTerje** [comment]
@asbjornu   Yes, agree, I think of bugfix branches just the same as feature branches, the naming just to make it easier to do the bookkeeping, so off develop it goes :-)

**Comment by CharliePoole** [comment]
@asbjornu 

> SemVer says it should be 3.4.1, unless it's a breaking change, which would make it 4.0.

Actually, not quite. Semver says it should be 3.4.1 if we release it by itself. If we release it mixed with new, non-breaking features, it should be 3.5.0. How we decide to do a release and what we group in it drives the decision.

In the past, the policy has been that a critical bug got fixed and forced a release, while other bugs got queued. However, we are generally flexible about this, so we are doing a forced release due to a critical bug, we have usually released all the other finished work along with it.

Right now, of course, we are throwing everything up in the air to see where it lands. :-)

Right now, with the NUnit project, we have enough overhead that releasing every three months seems like a reasonable goal. With the work on Cake and intro of GitVersion I think we can push that down to a month. That won't change how we do things much though. However, if releasing gets **extrememly** easy, we might release more even more rapidly, even to the point of one feature at a time, one critical bug at a time and something like weekly for non-critical bugs. That's my fantasy, anyway. :-)

**Comment by CharliePoole** [comment]
Both GitReleaseNotes and GitReleaseManager are available through Cake integration. We could look at them after we get a satisfactory GitVersion implementation up and running.

**Comment by asbjornu** [comment]
> Actually, not quite. Semver says it should be 3.4.1 if we release it by itself. If we release it mixed with new, non-breaking features, it should be 3.5.0. How we decide to do a release and what we group in it drives the decision.

Indeed. And as you say, the frequency of the releases will be a dominant factor for what version number the next release is going to be. Give it enough time and you're bound to have at least one new noteworthy feature in there, requiring you to do a minor increment. I'd like to point out that you can do large refactorings and such in feature branches that won't add a new feature to the end product and thus only requiring you to do a patch increment.

But, yeah, I think this is going to work itself out. Figuring out the version number for a release is always going to be a manual process, so that's not going to become much different regardless of the overall process and tools involved.

**Comment by CharliePoole** [comment]
So let's say we name branches as @asbjornu listed, with the addition of bugfix. His description of standard version numbers works for me except for some of the suffixes.

For example, why would you call the development branch alpha? Alpha has a traditional implication of a short trial period, leading to a beta period and then a final release. The develop branch, otoh, seems like a new thing in itself. FWIW, I would hope we can get away from alphas and betas in the future and either do an rc or just release.

On feature branches, are you considering just the word feature plus a number? That is, no designation of what the feature is about? How about nnn-Description, where nnn is the issue or pr number?

Release branches make more sense to me as rc versions rather than beta. In a beta, you will add more features while an rc is intended to use as is unless something goes wrong.

Feature branches as beta don't make sense either, especially if they are then merged into an alpha prefixed develop branch. Like develop, I think they are something new.

Support branches will probably go unused.

One question... is it important that all the suffixes we use be in a proper alphabetic order or only the suffixes that are used on the same branch?

**Comment by CharliePoole** [comment]
@nunit/core-team I'm somewhat blocked by not having a standard naming convention, so I'll lay out a proposal...

| Branch | Feed | Default Version | Proposed Version |
| --- | --- | --- | --- |
| `master` | NuGet | `3.4.1`, `3.5.0-beta1` | (same) |
| `develop` | MyGet | `3.5.0-unstable-nnnn` | `3.5.0-dev-nnnn` |
| `feature/*` | AppVeyor | `3.5.0-feature/SomeFeature` | `3.5.0-SomeFeature` |
| `bugfix/*` | AppVeyor | `3.4.3-bugfix/SomeBug` | `3.4.3-SomeBug` |
| `pull/*` | None | `3.4.1-PullRequest-nnnn` | `3.5.0-pr-nnnn` |
| `release/*` | AppVeyor | `3.5.0-beta-nnnn` | `3.5.0-rc-nnnn` |
| `hotfix/*` | AppVeyor | `3.4.1-beta-nnnn` | `3.4.1-rc-nnnn` |
| `support/*` | None | NA | NA |

**Note:** Not including PRs in the AppVeyor feed is a proposed change from what we now do.

**Comment by CharliePoole** [comment]
I should have mentioned one unstated assumption in my last comment... I'm assuming that the existence of a develop branch will lead us to phase out alphas and betas. We may want RCs, which are then released without change if all goes well, but the function of beta releases is pretty much taken over by the existence of the develop branch.

If we do phase out alphas and betas, then we may do it sooner on smaller projects and later on more complex ones like NUnit itself.

**Comment by CharliePoole** [comment]
I updated the proposed pre-release formats after trying the first set out on GitVersion. The updated values are more in line with what GitVersion expects.

Still waiting for some input on whether it's OK that these are only partially ordered. Specifically, the AppVeyor feed would continue to require selection of the exact build needed since there is no meaningful definition of "latest."

**Comment by OsirisTerje** [comment]
The version string suffixes should be in alphanumerical sequence, otherwise `*-branch` will be "Later" than `*-beta`, I am not sure that is what we want.  And *-dev will be later than both of these, so perhaps good to name packages created off the dev and feature branches as` *-alpha ?

**Comment by CharliePoole** [comment]
I added the GitVersion default values to the table, in order to clarify things a bit.

I understand the point about ordering pre-releases, but for the moment, I want to take the contrary view, just for the sake of discussion. I'm not totally wedded to this approach, but I'd like to see it explored before we decide. I was following two principles...
1. Ordering need only apply within a feed
2. The Appveyor CI feed doesn't have to be ordered

If we change either of those, then we have to change the suffixes. However, I would argue that a truly meaningful order is not possible across all feeds, so long as we are publishing work in process. I'll take this one feed at a time.

**NuGet**
We only put real public releases out to NuGet, which means and we only create those releases out to the master branch. This feed will always be in order.

**MyGet**
As proposed, we only put actual releases of the develop branch on MyGet. The GitFlow default used unstable as the tag and I used dev. I'd be happy to stick with unstable, which I suspect they chose because it's very high in the alphabet. However, all the releases on this feed would have the same tag, so it doesn't much matter.

If we wanted to call these beta releases, we could do that. But in that case, I would suggest putting them out to nuget as pre-releases, rather than using MyGet. In order to do that, we would have to stop doing the other sort of beta that we now do - i.e. a beta that appears right before the release is ready. IOW, we would be in a continuous beta state for the next release starting right after any master release.

**AppVeyor**
The AppVeyor feed is already unordered, at least in any meaningful way. That is, it's properly ordered because all the packages use the CI prefix and the appveyor build number, but the order has no meaning except that it's chronological. It includes different branches and PRs, built for different purposes. To use it, you have to cherry-pick the branch you want. This is OK since it is mostly used by us and sometimes by users testing out a particular bug. But it really has no useful "latest release" If you were to use it and install latest, you could as easily get a build of some proposed feature from a user as anything really meaningful.

Consequently, I haven't worried about keeping the appveyor build ordered in the proposal.

**Combining feeds**
The only meaningful combination of feeds is between NuGet and MyGet. That's because you may want to get the latest stable release of one component and the latest unstable of another. This will work because at any given time the NuGet feed has the latest build of the _current minor version_ of each component, while the MyGet feed has the _next minor version_.

---

Anyway, that's the logic that led me to do it as I did. I just want to be sure that logic is understood before we decide to go another way.

**Comment by CharliePoole** [comment]
Setting aside the labels used for the moment... I ran a series of 12 tests in this repo, creating fake features and hotfixes and merging them to see what happened. The base for all this was master at version 0.7.0. Here are the results, with a few notes attached. You can see the jobs and artifacts on AppVeyor.
1. Created feature/TestFeature1 branch, added a file and pushed it.
       Appveyor job:    0.7.1-TestFeature1.1+1.build.73
       Package version: 0.7.1-TestFeature1-1
2. Created features/TestFeature2 branch, added a file and pushed it.
       Appveyor job:    0.7.1-TestFeature2.1+1.build.74
       Package version: 0.7.1-TestFeature2-1
   Note: The 's' on features was an typo, but didn't matter
   because of the regex used to identify the branch.
3. Modified feature/TestFeature1 branch
       Appveyor job:    0.7.1-TestFeature1.1+2.build.75
       Package version: 0.7.1-TestFeature1-1
   NOTE: Expected 0.7.1-TestFeature1-2. Does this need fixing?
4. Merged feature 1 into develop
       Appveyor job:    0.8.0-dev.5.build.77
       Package version: 0.8.0-dev-5
   NOTE: It's '5' because I was fooling around in the branch earlier
5. Merged feature 2 into develop
       Appveyor job:    0.8.0-dev.7.build.78
       Package version: 0.8.0-dev-7
6. Created release-0.8.0 branch from develop, added changes.txt and pushed.
       Appveyor job:    0.8.0-rc.1+1.build.79
       Package version: 0.8.0-rc-1
7. Created a pull request for release-0.8.1 branch
       Appveyor job:    0.8.0-pr.10+9.build.80
       Package version: 0.8.0-pr-10
8. Used GitHub online interface to merge the request to master
       Appveyor job:    0.8.0+0.build.81
       Package version: 0.8.0
   NOTE: Earlier merges were done locally and then pushed
9. Created branch hotfix/0.8.1 from master, added a file and pushed
       Appveyor job:    0.8.1-rc.1+1.build.82
       Package version: 0.8.1-rc-1
10. Created a PR for the hotfix
        Appveyor job:    0.8.0-pr.11+2.build.83
        Package version: 0.8.0-pr-11
    NOTE: I thought this would be 0.8.1, but apparently not. Why?
11. Used GitHub online interface to merge the hotfix to master
        Appveyor job:    0.8.1+0.build.84
        Package version: 0.8.1
12. Manually merged the hotfix to develop
        Appveyor job:    0.9.0-dev.1.build.85
        Package version: 0.9.0-dev-1

---

The sequence needs review to ensure that incrementing is taking place as expected. Some specific issues are noted on the affected item.

The only publishing that is currently being done is for master. I think deployment has to be phased in after we convert to using GitVersion. The 0.8.0 and 0.8.1 master builds were deployed on MyGet.

**Comment by ChrisMaddock** [comment]
Apologies for referring to something a few days old...

> Note: Not including PRs in the AppVeyor feed is a proposed change from what we now do.

Does this potentially limit how we can work on the core project, or is there a workaround I've missed?

My situation: I'm currently working on a framework change which will require a couple of new LINQ additions. To do that, I want to add them to this assembly, and get a 'temporary' package that I can base my framework work off. (From a nuget feed rather than a local package, so I can submit a working [WIP] PR to the framework.)

My only way to currently do that seems to be to submit a [DoNotMerge] PR to this repo, and then use the package from the appveyor feed. If we don't have PR packages, would there be another way around this problem?

This specific case could potentially be solved by teams having branch access to all dependency repo's, but would that then equally limit a non-team members ability to contribute?

**Comment by rprouse** [comment]
@ChrisMaddock that is a good catch on your part, I missed it. I am using that ability constantly working on the .NET Core projects at the moment. I make a fix in one repo, as you said, create a PR, then pull in the NuGet package in the second repo and test or use the changes.

We are adding a single NUnit wide AppVeyor NuGet feed. I think it is fine that PR builds are not pushed to it, but we could work around the problem by continuing to maintain project specific feeds that are only used for development. @CharliePoole was this the intention?

**Comment by CharliePoole** [comment]
@rprouse @ChrisMaddock 

This question is quite a can of worms. :-)

@ChrisMaddock From your post I take it that you are OK doing local testing but you want to be able to submit a PR that will build on AppVeyor. So, for example, you may be working on a change in the framework that depends on some new class being in NUnit.System.Linq. Have I got that right?

Further, I take it that you are already set up with a local feed where you can test things before submitting.

So..., this introduces quite a few issues... :-)
1. The current AppVeyor feed cannot be used that way unless you were to submit the PR for the framework project specifying a _specific_ build in packages.config, e.g.: `version=[0.9.0-CI-xxx-PR-yyy]`. That's because the feed, while ordered, is not ordered in any meaninful way. If you just let it get the "latest" then it could be superseded by somebody else's branch or PR build that didn't include your change. If you did use the exact version in the PR, then you would have to remember to change it back to a correct version before merging, which seems risky.
2. The way we envisioned making changes across different repositories - and the way I have been doing it for the NUnit 3 Test Adapter - is that the underlying change has to be made first, merged and published to either MyGet or Nuget. It can be a pre-release but it has to be in ordered sequence on the same branch - master right now, develop when we implement it. This is harder than having everything in one project and we knew that when we decided to split things up. But we decided to accept that obstacle with the idea that only a minority of changes would involve changing some other package before it could be made. That seems to be true once a package is developed. It's less true for completely new things like the .NET Core runner or (in the recent past) the NUnit 3 Test Adapter. We hope that the benefit we get from separation will outweigh the difficulties.
3. For the specific change you want to make, you should submit an issue on this project. Explain what you want to do and why. We will review that and tell you what we think. Assuming the change is accepted as an issue, then do a PR and get it published. Then go on with the change you want to make. The alternatives we will consider to changing NUnit.System.Linq are (1) excluding the new feature (I assume it's a framework feature) from the 2.0 build or (2) asking you to implement it without use of System.Linq.
4. I can't be sure what you are working on, but *_if *_ it's the issue about expressing properties as lambdas, my inclination would be to exclude it from the 2.0 build. If a user has decided to write tests using .NET 2.0 rather than 3.5, then they probably aren't expecting to use lambdas in the interface with NUnit. They always have the option to target 3.5 in the tests, even if the SUT targets 2.0. In general, I think we can use (and do) System.Linq features in our implementation of the 2.0 build but should keep them out of the user interfaces.
5. Back to the more general question that underlies this IMO: "What is each feed for?" Here's my take on it...
   - **NuGet** is for real production releases to users, including any alphas, betas or rcs that we actually intend for users to download and try out.
   - **MyGet** is for our own internal use where one of our packages depends on another and for users who want to get a continuous development feed. It actually supports multiple feeds if we find the need to have them and we could set one up that way as Rob suggests. So far, we just have one feed.
   - **AppVeyor** is for testing, generally by us but sometimes by users who select a very specific package whose name we give them. Unless we were to restrict it to master builds only, its contents are mixed and "latest" really has no meaning.

This is pretty important stuff with a lot depending on it, so please argue with me if what I have written doesn't seem right. We **want** to get this right!

**Comment by CharliePoole** [comment]
I'll add one more thing and then wait for comments. :-)

**If** somebody is working on a feature that requires constant changes to two independent packages, going back and forth between them. The approach of completing the lower level package and publishing it first doesn't work. This has been the case with the gui, for example, where I constantly discovered new things I wanted out of the engine.

The way I dealt with that was to pull the engine project into my solution locally and work on both. But before I could publish anything for the gui, I had to first make the underlying changes to the engine and get them merged. It's not convenient, as already stated.

**If** we were happy to continue all our development with a small number of devs, as now, then we could have kept everything together in one big project. We are taking a chance that separate, smaller projects will help us pull in more devs.

**Comment by rprouse** [comment]
@CharliePoole I think this is a case of constant changes in two independent packages. Right now, that is happening between `nunit.portable.driver` and `dotnet-test-nunit`. I have currently solved the problem by adding the AppVeyor nuget feed for the portable driver to the dotnet test project and consuming specific builds of PRs. It means a slight delay waiting for the AppVeyor build, but it is manageable.

Another workaround is to package on the developer's local machine and add the packages to a local NuGet feed directory. I do that for faster turnarounds. We could work that way, but we should document it.

**Comment by ChrisMaddock** [comment]
@CharliePoole: Can of worms indeed! Your assumptions about what I'm doing are all correct, for context, the issue is nunit/nunit#26. I'll comment point by point below! 

> The current AppVeyor feed cannot be used that way unless you were to submit the PR for the framework project specifying a specific build in packages.config, e.g.: version=[0.9.0-CI-xxx-PR-yyy]. [...] If you did use the exact version in the PR, then you would have to remember to change it back to a correct version before merging, **which seems risky.**

Yes - far from ideal! Risky, and requires enforced additional to-ing and fro-ing - if the person who reviews is happy with everything as it is, they can't just merge as further changes will still be required.

> The way we envisioned making changes across different repositories [...] is that the underlying change has to be made first, merged and published to either MyGet or Nuget. [...]
> 
> For the specific change you want to make, you should submit an issue on this project [...]
> 
> I can't be sure what you are working on, but *if \* it's the issue about expressing properties as lambdas

I think these are all valid points for the issue at hand, but perhaps we should be looking at this more with multiple dependencies more generally. NUnit.Linq is a relatively simple example, in that we have a relatively clear idea of the specific missing namespaces required for the implementation.

From the point of view of how _I personally_ would work, it feels like it would slow changes down significantly if I had to wait for a review/merge mid-way through a substantial change. Of course, I could do this all locally, but it feels messy in terms of replacing a NuGet reference with a local one.

The 'least-worst' solution to me, feels like being able to just modify the packages.config to an appveyor-esque version number. The change should be relatively clear to catch in review. Could we potentially also set up a CI task which only uses MyGet/NuGet package restore sources, so a PR _couldn't_ be merged with Appveyor packages in the config?

**Comment by CharliePoole** [comment]
@rprouse Yes, I use a local repo myself now when working on the test adapter. But that's basically past the "back and forth" stage now. Since all the projects are on my machine and adding projects to a solution doesn't copy them, I find a master solution to be fastest in that situation. And it avoids my creating packages for changes that are more of a "what if" nature and may be abandoned later.

I think the way you are working now - using the AppVeyor feed - works well because you are the only one committing to those packages. It might be a problem if you were depending on the engine, for example.

**Comment by CharliePoole** [comment]
@ChrisMaddock Yes, NUnit.System.Linq is both a simple example and is also completely different from all other packages in that it is only published for our own use. Additionally, it's not our code so we didn't expect to change it once created. We made it a package rather than keeping it in the project because we expect to use it in our engine build, which is about to be separated.

I think we have two entirely separate use cases to deal with:
1. New development of a kind that implies entirely new features in the referenced package - the "back and forth" development that Rob and I are talking about. In such cases, you may not even know in advance exactly what features you want in the lower-level package.
2. More "normal" development where you need an extra feature in a well-established and stable package. Usually, you can specify in advance what that feature is and if not you can experiment locally until you are ready to specify it. I think this is probably the case you are dealing  with.

I think the elephant in the room here is that we actually **want** to slow down some kinds of changes. If a package is being depended on by several different using packages, then we need to be sure a change for the benefit of one of them won't mess up the others. That's what the cycle of issue-discuss-develop-merge is all about. When needful, we can do it fairly quickly, but of course, it's usually a matter of days, not hours, unless you and the reviewer are in just the right time zone relation to one another.

To put it in perspective, here are some changes to the test adapter that necessitated (and waited for) engine or framework changes:
- Getting accurate file and line number locations in the adapter required adding fields to the representation of a test in the framework.
- Displaying immediate output in the Output window required making the framework actually produce such output in the first place, rather than bundling it with the Test Result, and changing the engine to allow passing the output on.
- Recognizing special error conditions in the adapter required adding new custom exceptions to the engine.

Each of these items resulted in two separate issues, separate coding and tests and separate merging. It definitely would be faster if we didn't have to do all this, but with an application in use by hundreds of thousands of developers, I've always felt that a level of care required.

In particular, the creation of an issue for each item is aimed at actually saving you time. If it turns out that what you want to do isn't supported by everyone else, it feels much more efficient to have you type in a paragraph's worth of info as compared to writing the code first. I'll grant you that this doesn't always _feel_ more efficient to a developer who would rather be coding than explaining. :-)

Others may have different views. Here's mine...
1. We want to retain a level of control and care over the packages we depend on that is even higher than what we demand of ourselves for the top-level packages, because they can affect more things.
2. So long as we can meet the standards of item 1, then we want to make the process as efficient as possible. I'd be in favor of creating some feed to which you could commit your changes and then use them if we could figure out a way to trigger use of that feed in our CI build based on some info we pass to it. That implies something that comes from GitHub, like the type of build or the name of the branch or a tag that you add... We don't want to change any files if we can avoid it because we would have to change them back.

**Comment by CharliePoole** [comment]
Back to the original question, after a wide digression. @ChrisMaddock Why do you prefer to use the PR build over the branch build? The reason that I suggested dropping PR builds is that it didn't seem to have any particular use for testing purposes. Maybe I missed something?

**Comment by ChrisMaddock** [comment]
> @ChrisMaddock Why do you prefer to use the PR build over the branch build?

The core team doesn't (currently) have branch access to this repository. That was what I meant in my original post by a team having branch access to all dependency repos - although that still wouldn't solve the problem for non-team contributors, which it feels like we should also be able to support.

I agree there's value in slowing down large changes. However, it feels like the additions I would want to make in NUnit.Linq would be best reviewed in context of the 'actual' changes I wanted to make in the Framework. It may be that with review, we decide the change the framework implementation, and no longer need the NUnit.Linq changes, or alternatively need further additions.

(Mostly unrelated, I think you make a good point that this particular feature is best just excluded from .NET 2.0 - but I'll move that discussion over to the original issue!)

**Comment by CharliePoole** [comment]
@ChrisMaddock AH! Well that's just an artifact of the current state of flux of our team setup, particularly as regards to what NUnit.Core means now and will eventually mean. It's a matter of about 60 seconds for me to give you write access to NUnit.System.Linq, completely apart from the team structure.

Contributors are a different matter, but it's hard to imagine a contributor starting off with something so complex that they needed to do multiple package updates. If it happened and they seemed credible candidates, we would probably make them team members of some kind.

NUnit.System.Linq is an exception in all this. It's not assigned to a team because we don't want people to start writing code for it. Rather, we would import additional types if needed.

I'll join you on #26 for more discussion. :-)

**Comment by ChrisMaddock** [comment]
Ok, sure.

So, to play devil's advocate, if using branch packages IS the right way to approach this problem, then would there be any issue also pushing PR packages to the same feed?

It would be rare for someone to need one, but if it's just a matter of which CI webhooks are enabled, seems worthwhile for the odd occasion they're useful?

**Comment by CharliePoole** [comment]
Well, it would depend on the purpose. My idea of dropping PR builds from the AppVeyor feed was (1) to simplify it, since it's one of three different kinds of things mixed in the same feed, and (2) because I can't figure out what a PR is particularly good for. 

When you push a change to a pull request, you get two builds published, one for the branch and one for the PR. If it's approved, you then get a third build, this time of master. If you do it fast enough, it's probably identical to the PR build.

I'm thinking that we probably don't want it published anywhere if it's not approved. And if it's approved you can get that master build, which appears on both AppVeyor and MyGet. Is there a use case where you want to get the PR build rather than the branch before we have decided whether to merge it?

I'm not opposed to it but I am against doing it just in case we might need it. I'd like to have some idea of what we may need it for.

**Comment by ChrisMaddock** [comment]
The only use case I was thinking of is when a user doesn't have branch access. In that case, waiting for review/merge before being able to move on isn't ideal.

I agree it sounds like it would be a rare situation however - it's just a question whether it's worth preventing that, just to simplify the Appveyor feed.

**Comment by asbjornu** [comment]
Sorry for being away for so long. I'm on vacation for 4 weeks now, but are a bit online now and then. Is there anything you need from me in this thread or are we just down to the discussion of which package feeds we should have, now? Please let me know if there's anything I need to do as I don't have time to read the entire discussion until august. 😄

**Comment by CharliePoole** [comment]
@asbjornu Thanks for letting us know. You're right, I think it's now a matter of our deciding what we want rather than how to do it.
